File: webapp/lib/bait.rb
# frozen_string_literal: true

# The set of valid bait types and validation of a bait name.
# Bait types are exactly: glory, riches, undead, power.
module Bait
  GLORY  = :glory
  RICHES = :riches
  UNDEAD = :undead
  POWER  = :power

  ALL = [GLORY, RICHES, UNDEAD, POWER].freeze

  module_function

  # Normalize an arbitrary value (string or symbol) to a known bait symbol,
  # raising if it is not one of the four valid types.
  def normalize(value)
    bait = value.to_s.downcase.to_sym
    return bait if ALL.include?(bait)

    raise ArgumentError, "unknown bait type: #{value.inspect}"
  end

  def valid?(value)
    ALL.include?(value.to_s.downcase.to_sym)
  end
end

File: android/app/src/main/java/com/dungeonboss/model/Bait.kt
package com.dungeonboss.model

/**
 * The set of valid bait types and validation of a bait name.
 * Bait types are exactly: glory, riches, undead, arcane.
 */
enum class Bait {
    GLORY, RICHES, UNDEAD, ARCANE;

    companion object {
        /**
         * Normalize an arbitrary value (string, symbol, or a [Bait]) to a known
         * bait, raising if it is not one of the four valid types.
         */
        fun normalize(value: Any?): Bait {
            val name = value.toString().lowercase()
            return entries.firstOrNull { it.name.lowercase() == name }
                ?: throw IllegalArgumentException("unknown bait type: $value")
        }

        fun isValid(value: Any?): Boolean =
            entries.any { it.name.lowercase() == value.toString().lowercase() }
    }
}
